src/magic_folder/test/test_snapshot.py

```diff
@@ -0,0 +1,118 @@
+import io
+import os
+
+from twisted.internet import defer
+from twisted.python.filepath import (
+    FilePath,
+)
+from twisted.web.resource import (
+    Resource,
+)
+from twisted.web.client import (
+    Agent,
+    FileBodyProducer,
+)
+
+from treq.client import (
+    HTTPClient,
+)
+from treq.testing import (
+    RequestTraversalAgent,
+    RequestSequence,
+    StubTreq,
+)
+
+from testtools.matchers import (
+    StartsWith,
+)
+
+from magic_folder.snapshot import create_snapshot
+
+from .fixtures import (
+    NodeDirectory,
+)
+from .common import (
+    ShouldFailMixin,
+    SyncTestCase,
+    AsyncTestCase,
+    skipIf,
+)
+
+
+class _FakeTahoeRoot(Resource):
+    """
+    This is a sketch of how an in-memory 'fake' of a Tahoe
+    WebUI. Ultimately, this will live in Tahoe
+    """
+
+
+class _FakeTahoeUriHandler(Resource):
+    """
+    """
+
+    isLeaf = True
+
+    def render(self, request):
+        return b"URI:DIR2-CHK:some capability"
+
+
+def create_fake_tahoe_root():
+    """
+    Probably should take some params to control what this fake does:
+    return errors, pre-populate capabilities, ...
+    """
+    root = _FakeTahoeRoot()
+    root.putChild(
+        b"uri",
+        _FakeTahoeUriHandler(),
+    )
+    return root
+
+
+
+class TahoeSnapshotTest(AsyncTestCase):
+    """
+    Tests for the snapshots
+    """
+
+    @defer.inlineCallbacks
+    def setUp(self):
+        """
+        Create a Tahoe-LAFS node which contain some magic-folder configuration
+        and run it.
+        """
+        yield super(TahoeSnapshotTest, self).setUp()
+        self._nodedir = self.useFixture(
+            NodeDirectory(
+                path=FilePath(self.mktemp()),
+            )
+        )
+        self._root = create_fake_tahoe_root()
+        self._agent = RequestTraversalAgent(self._root)
+        self._client = HTTPClient(
+            self._agent,
+            data_to_body_producer=_SynchronousProducer,
```

Comment:
do we need to include `_SynchronousProducer` in the import list of `treq.testing`?

Reply:
I couldn't find a way to do it without that .. I was copying what's in Treq's code (for `StubTreq` .. which could be used instead, maybe, but you'd also need to rely on private things to get at the agent then).

Without that, passing anything but `None` to PUT or POST requests hangs.

Reply:
`data_to_body_producer` seems to be clearly public.  It's only the `_SynchronousProducer` part of this that is a problem.  We could just copy that into magic-folder (or tahoe-lafs eventually).  Or, looks like I did implement my own in txkube, so we could copy that one instead.